File: services/api-node/src/middleware/metrics.js
import { httpRequestsTotal, httpRequestDuration } from '../metrics.js';

export function metricsMiddleware(req, res, next) {
  const start = process.hrtime();

  res.on('finish', () => {
    const diff = process.hrtime(start);
    const duration = diff[0] + diff[1] / 1e9;
    
    // Use req.path or req.originalUrl as fallback
    const path = req.route?.path || req.path || req.originalUrl || 'unknown';
    const status = res.statusCode.toString();

    httpRequestsTotal
      .labels(req.method, path, status)
      .inc();

    httpRequestDuration
      .labels(req.method, path, status)
      .observe(duration);
  });

  next();
}

File: services/core-go/internal/middleware/metrics.go
package middleware

import (
	"net/http"
	"time"
	"core-go/internal/metrics"
)

type statusRecorder struct {
	http.ResponseWriter
	status int
	wroteHeader bool
}

func (r *statusRecorder) WriteHeader(code int) {
	if !r.wroteHeader {
		r.status = code
		r.wroteHeader = true
		r.ResponseWriter.WriteHeader(code)
	}
}

func (r *statusRecorder) Write(b []byte) (int, error) {
	if !r.wroteHeader {
		r.WriteHeader(http.StatusOK)
	}
	return r.ResponseWriter.Write(b)
}

func Metrics(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		start := time.Now()

		recorder := &statusRecorder{
			ResponseWriter: w,
			status:         200,
			wroteHeader:    false,
		}

		next.ServeHTTP(recorder, r)

		duration := time.Since(start).Seconds()

		metrics.HttpRequestsTotal.WithLabelValues(
			r.Method,
			r.URL.Path,
			http.StatusText(recorder.status),
		).Inc()

		metrics.HttpRequestDuration.WithLabelValues(
			r.Method,
			r.URL.Path,
		).Observe(duration)
	})
}
